error.go

```diff
@@ -20,24 +24,110 @@ import "encoding/json"
 //     }
 //
 type Error struct {
-	Desc string `json:"desc"`
-	Docs string `json:"docs"`
-	Kind string `json:"kind"`
+	Cause error   `json:"cause,omitempty"`
+	Desc  string  `json:"desc,omitempty"`
+	Docs  string  `json:"docs,omitempty"`
+	Kind  string  `json:"kind,omitempty"`
+	Stack []Stack `json:"stack,omitempty"`
 }
 
-func (e *Error) Error() string {
-	return toStringCase(e.Kind)
+type Stack struct {
+	File string `json:"file,omitempty"`
+	Line int    `json:"line,omitempty"`
 }
 
-func (e *Error) GoString() string {
-	return e.String()
+func (e *Error) Error() string {
+	if e.Cause != nil {
+		return e.Cause.Error()
+	}
+
+	k := toStringCase(e.Kind)
+
+	if e.Desc != "" {
+		return fmt.Sprintf("%s: %s", k, e.Desc)
+	}
+
+	return k
 }
 
-func (e *Error) String() string {
-	b, err := json.Marshal(e)
+func (e *Error) MarshalJSON() ([]byte, error) {
+	type ErrorClone Error
```

Comment:
Why is this type needed?

Reply:
That is some nasty `microerror` and `micrologger` implementation detail. Our logger uses [the go-kit logger](https://github.com/go-kit/kit/blob/10f464fb11741e507cbb5201eb889c37dd51dbab/log/json_logger.go#L37-L61). There are some type assertions defining how to represent the original `interface{}` and its actual underlying type. When the provided type, which is our `*microerror.Error` implements the JSON marshaler it properly decodes the structure in the resulting JSON log. When you implement the marshaler yourself, you cannot marshal the same type using the standard lib. The clone type prevents a deadlock. We can mention all this in the comment to make it clear later on once it becomes relevant again eventually.